### ISSUE TITLE
Linking Qt6 application

### ISSUE THREAD

**Issue by slorenzen** [mxe/mxe#2951]
Dear forum,

I am trying to compile a Hello-World-Qt6 application using MXE. Compilation works fine, but linking does not work:

```
[ 40%] Linking CXX executable muell.exe
/home/stephan/mxe/usr/i686-w64-mingw32.static/qt6/lib/libQt6Gui.a(qplatformdialoghelper.cpp.obj): duplicate section `.data$_ZZN9QMetaType23registerMutableViewImplI5QListI4QUrlE9QIterableI13QMetaSequenceEEEbSt8functionIFbPvS8_EES_S_E10unregister[__ZZN9QMetaType23registerMutableViewImplI5QListI4QUrlE9QIterableI13QMetaSequenceEEEbSt8functionIFbPvS8_EES_S_E10unregister]' has different size
```
[...many more of these errors...]
```
/home/stephan/mxe/usr/bin/i686-w64-mingw32.static-ld: /home/stephan/mxe/usr/lib/gcc/i686-w64-mingw32.static/11.3.0/../../../../i686-w64-mingw32.static/lib/../lib/libwebp.a(libwebpencode_la-picture_csp_enc.o): in function `ImportYUVAFromRGBA':
/home/stephan/mxe/tmp-libwebp-i686-w64-mingw32.static/libwebp-1.3.0/src/enc/picture_csp_enc.c:508: undefined reference to `SharpYuvInit'
/home/stephan/mxe/usr/bin/i686-w64-mingw32.static-ld: /home/stephan/mxe/usr/lib/gcc/i686-w64-mingw32.static/11.3.0/../../../../i686-w64-mingw32.static/lib/../lib/libwebp.a(libwebpencode_la-picture_csp_enc.o): in function `PreprocessARGB':
/home/stephan/mxe/tmp-libwebp-i686-w64-mingw32.static/libwebp-1.3.0/src/enc/picture_csp_enc.c:180: undefined reference to `SharpYuvGetConversionMatrix'
/home/stephan/mxe/usr/bin/i686-w64-mingw32.static-ld: /home/stephan/mxe/tmp-libwebp-i686-w64-mingw32.static/libwebp-1.3.0/src/enc/picture_csp_enc.c:180: undefined reference to `SharpYuvConvert'
/home/stephan/mxe/usr/bin/i686-w64-mingw32.static-ld: /home/stephan/mxe/usr/lib/gcc/i686-w64-mingw32.static/11.3.0/../../../../i686-w64-mingw32.static/lib/../lib/libmingw32.a(lib32_libmingw32_a-crt0_c.o): in function `main':
/home/stephan/mxe/tmp-gcc-i686-w64-mingw32.static/gcc-11.3.0.build_.crt/../gcc-11.3.0.build_/mingw-w64-v10.0.0/mingw-w64-crt/crt/crt0_c.c:18: undefined reference to `WinMain@16'
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/muell.dir/build.make:226: muell.exe] Fehler 1
make[1]: *** [CMakeFiles/Makefile2:84: CMakeFiles/muell.dir/all] Fehler 2
make: *** [Makefile:136: all] Fehler 2
```

I googled a lot for these errors. According to one post stating that compiler options not matching between the Qt build and my application build, I looked in the build files of Qt6Gui and set the same options (besides "-c") in my CMakeLists.txt:
```
set(CMAKE_CXX_FLAGS "\
	-pipe -fno-keep-inline-dllexport -O2 -Wall -Wextra -Wextra -fexceptions -mthreads \
	-DUNICODE -D_UNICODE -DWIN32 -DMINGW_HAS_SECURE_API=1 \
	-DQT_NO_DEBUG -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_SQL_LIB -DQT_CORE_LIB -DQT_NEEDS_QMAIN \
	-I/home/stephan/mxe/usr/i686-w64-mingw32.static/qt6/include \
	-I/home/stephan/mxe/usr/i686-w64-mingw32.static/qt6/include/QtWidgets \
	-I/home/stephan/mxe/usr/i686-w64-mingw32.static/qt6/include/QtGui \
")
set(Qt6Widgets_DIR "/home/stephan/mxe/usr/i686-w64-mingw32.static/qt6/lib/cmake/Qt6Widgets")
```

Are those duplicate sections warnings (which I might ignore) or errors?

The undefined references to SharpYuv stuff appeared before when I compiled MXE, which was the reason why I had to exclude several files from the MXE build (among others, qt6-qttools). However, MXE did build Qt6Widgets without errors. Do I have to repair these SharpYuv errors or is there a complier option to skip these dependencies?

The last line in the error output stating an undefined reference to WinMain seems really disturbing. How can I fix that one?

I am using the latest version of MXE (downloaded with git just today), my platform is 
```
stephan@stephtop:~/mxe$ uname -a
Linux stephtop 6.0.0-6-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.0.12-1 (2022-12-09) x86_64 GNU/Linux
```

Thanks for any advise!

**Comment by hr-98** [comment]
I'm having issues with SharpYuv missing too and can't get compile qt6 tools.

**Comment by mbunkus** [comment]
If I understand what I've read on the web correctly, `libsharpyuv` used to be bundled with `libwebp` and isn't anymore, or they're in the process of unbundling.

As a matter of fact the source for `libwebp` 1.3.0 still contains `libsharpyuv`, and that library is indeed built & installed properly. Additionally it looks like the `pkg-config` file for it does still list it as a library to link:

```
[0 mosu@sweet-chili (master-20230208-8e18ff2d) /opt/mxe] x86_64-w64-mingw32.static-pkg-config --static --libs libwebp
-lwebp -lsharpyuv -lm
```

The corresponding `pkg-config` file contains `libshapyuv` as a requirement:

```
[0 mosu@sweet-chili (master-20230208-8e18ff2d) /opt/mxe] rg -i 'libs.private|sharpyuv' /opt/mxe/usr/x86_64-w64-mingw32.static/lib/pkgconfig/libwebp.pc
9:Requires: libsharpyuv
12:Libs.private: -lm
```

And here's where I find things get strange: `libshapyuv.pc` only contains its own library in `Libs` but not `Libs.private`:

```
[0 mosu@sweet-chili (master-20230208-8e18ff2d) /opt/mxe] rg -i 'libs.private|sharpyuv' /opt/mxe/usr/x86_64-w64-mingw32.static/lib/pkgconfig/libsharpyuv.pc
6:Name: libsharpyuv
10:Libs: -L${libdir} -lsharpyuv
11:Libs.private: -lm
```

I don't know enough about `pkg-config` to say whether this is correct or not.

BTW, I suggest re-titling this issue to something like "qttools & qt6-qttools fail to build due to not likinging against libsharpyuv`.

**Comment by mbunkus** [comment]
As a temporary workaround reverting 2dd7a867b150518e9fc5497f1060d6734843cf12 and with that `libwebp` from 1.3.0 back to 1.2.4 works; Qt 6.4.2 builds fine with it.

**Comment by hr-98** [comment]
Thanks, that works for me

**Comment by nostar** [comment]
I also reverted back to libwebp 1.2.4 and was able to build qt6, but when I run qmake on my qt app that uses multimedia and serialport, I get the error
Project ERROR: Unknown module(s) in QT: multimedia serialport

I also built qt5 and it works fine.

I confirmed that the qt6-qtserialport and qt6-qtmultimedia packages were downloaded and built during the make process.

**Comment by mabrand** [comment]
The linking problem related to libsharpyuv is probably solved and/or worked around now. Could you please check again?

**Comment by mbunkus** [comment]
I can confirm that building current `master` works fine for me, including building my app with it afterwards. Thanks!

**Comment by mbunkus** [comment]
(posted too soon a minute ago)

I can confirm that MXE itself builds fine now with the current `master`, but my own application fails at the linking stage the same undefined references to symbols `SharpYuvInit` & `SharpYuvGetConversionMatrix`. I don't have that much time to dive into it at the moment, but I'll try to dig deeper.

Do you have any pointers what those workarounds you've mentioned were? They might help me as well. Thanks.

**Comment by mabrand** [comment]
Yes, take a look at d8bfdc137d7c01e0458861303b25ecd1d7cfc5e0 and d8bfdc137d7c01e0458861303b25ecd1d7cfc5e0.

**Comment by mbunkus** [comment]
Thing is, I'm not linking against `-lwebm` explicitly, neither against `-lsharpyuv`. I'm calling `qmake` to produce the `Makefile`. That `Makefile` only contains `-lwebm` and two other `-lwebm…` libraries, but no `-lsharpyuv`. Therefore linking fails with symbols not being found. It isn't the order that's wrong, it's the fact that the library isn't linked against at all.

This is on `x86_64-w64-mingw32.static`.

It's actually easy to reproduce:

1. Create empty `testme.h` & `testme.cpp`
2. Create the `testme.qmake` from below
3. Run `x86_64-w64-mingw32.static-qt6-qmake testme.qmake`
4. Run `make`

Even though there's no code at all, linking fails. Adding `-lsharpyuv` do the linker command works.

`testme.qmake`:

```
HEADERS = testme.h
SOURCES = testme.cpp
```

In other words, the configuration of Qt6's webp plugin is wrong wrt. linking, at least in static situations. Haven't tried non-static.

**Comment by mbunkus** [comment]
OK, I just gave it a full try again. Here's how I can reproduce it:

1. Checkout a clean revision 32d1381a426aac8dd07ae32c66e7d94d4cb97eb3
2. Create `settings.sh` ; content see below
3. Remove `usr` if it still exists & create a new, empty one
4. compile with `make gettext libiconv zlib boost flac ogg pthreads vorbis cmark libdvdread gmp qt6 qt6-qtmultimedia` (these packages are what I need for my application)
5. in an empty directory, create empty `testme.h` & `testme.cpp`
6. in the same directory create `testme.qmake` ; content see below
7. run `x86_64-w64-mingw32.static-qt6-qmake testme.qmake`
8. run `make`

Yes, it'll complain about linker errors wrt. `main`, but more importantly it also complains about linker errors referring to `SharpYuvInit` & others.

[Here](https://gist.github.com/mbunkus/25121ce61a8611d3bf76b3e6ac2a9226) is a gist of the full run.

`settings.mk`:

```
MXE_USE_CCACHE =
MXE_PLUGIN_DIRS += plugins/gcc12
MXE_TARGETS=x86_64-w64-mingw32.static
JOBS=33
```

`testme.qmake`:

```
SOURCES=testme.cpp
HEADERS=testme.h
```

**Comment by mabrand** [comment]
qt6-qtimageformats looks like the culprit.
f5037ee75c3779c25544f76182d6412fd06bab09 should fix it.

**Comment by mbunkus** [comment]
Can confirm! With that revision my program builds & links just fine. Thank you so much!
